google_guest_agent/network/systemd_networkd.go

```diff
@@ -0,0 +1,107 @@
+package network
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/guest-agent/google_guest_agent/run"
+	"github.com/GoogleCloudPlatform/guest-agent/metadata"
+	"github.com/GoogleCloudPlatform/guest-logging-go/logger"
+)
+
+const configDir = "/etc/systemd/network"
+
+type systemdnetworkd struct{}
+
+func (n systemdnetworkd) IsManaging(ctx context.Context, iface string) (bool, error) {
+	// First check if the service is running.
+	res := run.WithOutput(ctx, "/bin/sh", "-c", "systemctl status systemd-networkd.service | grep Active:")
+	if res.ExitCode != 0 {
+		return false, fmt.Errorf("failed to get status of systemd-networkd service: %v", res.StdErr)
+	}
+	if !strings.Contains(res.StdOut, "running") {
+		// Service is not running, so it can't be managing an interface.
+		return false, nil
+	}
+
+	// Check systemd network configuration.
+	res = run.WithOutput(ctx, "/bin/sh", "-c", "networkctl status | grep "+iface)
+	if res.ExitCode != 0 {
+		return false, fmt.Errorf("failed to check systemd-networkd network status: %v", res.StdErr)
+	}
+	if strings.Contains(res.StdOut, "configured") {
+		return true, nil
+	}
+	return false, nil
+}
+
+func (n systemdnetworkd) Setup(ctx context.Context, payload []metadata.NetworkInterfaces) error {
+	// Create a network configuration file with default configurations for each network interface.
+	ifaces, err := interfaces(payload)
+	if err != nil {
+		return err
+	}
+	for _, iface := range ifaces {
+		logger.Debugf("write network config for %s", iface)
+
+		var config *os.File
+
+		// The number in beginning of the file name signifies its priority. We want this
+		// to take priority before any of the default configurations that systemd-networkd
+		// has set.
+		config, err = os.Create(fmt.Sprintf("%s/50-%s.network", configDir, iface))
+		if err != nil {
+			return err
+		}
+		defer config.Close()
+
+		// Default, simple configuration for any network interface.
+		contents := []string{
+			managedKey,
+			"[Match]",
+			fmt.Sprintf("Name=%s", iface),
+			"",
+			"[Network]",
+			"DHCP=yes",
+			"IPForward=true",
+		}
+		_, err = config.WriteString(strings.Join(contents, "\n"))
+		if err != nil {
+			return err
+		}
+	}
+	// Restart the systemd-networkd service.
+	logger.Debugf("restarting systemd-networkd")
+	return run.Quiet(ctx, "systemctl", "restart", "systemd-networkd.service")
+}
+
+func (n systemdnetworkd) Rollback(ctx context.Context, payload []metadata.NetworkInterfaces) error {
+	// Delete the configuration files created by the agent.
+	// Read the directory.
+	f, err := os.Open(configDir)
+	if err != nil {
+		return fmt.Errorf("failed to open systemd-networkd configuration directory: %v", err)
+	}
+	files, err := f.Readdir(-1)
+	f.Close()
```

Comment:
you're missing error check on this `close()`

Reply:
Done.